modules/infinite-scroll/infinity.php

```diff
@@ -1679,6 +1692,182 @@ public function add_mejs_config( $scripts_data ) {
 		}
 		return $scripts_data;
 	}
+
+	/**
+	 * Load AMP specific hooks.
+	 *
+	 * @return void
+	 */
+	public function amp_load_hooks() {
+		if ( class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ) {
+			$template = self::get_settings()->render;
+
+			add_filter( 'jetpack_infinite_scroll_load_scripts_and_styles', '__return_false' );
+
+			add_action( 'template_redirect', array( $this, 'amp_start_output_buffering' ), 0 );
+			add_action( 'shutdown', array( $this, 'amp_output_buffer' ), 1 );
+
+			if ( is_callable( "amp_{$template}_hooks" ) ) {
+				call_user_func( "amp_{$template}_hooks" );
+			}
+
+			// Warms up the amp next page markup.
+			// This should be done outside the output buffering callback started in the template_redirect.
+			$this->amp_get_footer_template();
+		}
+	}
+
+	/**
+	 * Start the AMP output buffering.
+	 *
+	 * @return void
+	 */
+	public function amp_start_output_buffering() {
+		ob_start( array( $this, 'amp_finish_output_buffering' ) );
+	}
+
+	/**
+	 * Flush the AMP output buffer.
+	 *
+	 * @return void
+	 */
+	public function amp_output_buffer() {
+		if ( ob_get_contents() ) {
+			ob_end_flush();
+		}
+	}
+
+	/**
+	 * Filter the AMP output buffer contents.
+	 *
+	 * @param string $buffer Contents of the output buffer.
+	 *
+	 * @return string|false
+	 */
+	public function amp_finish_output_buffering( $buffer ) {
+		// Hide WordPress admin bar on next page load.
+		$buffer = preg_replace(
+			'/id="wpadminbar"/',
+			'$0 next-page-hide',
+			$buffer
+		);
+
+		// Get the theme footers.
+		$footers = apply_filters( 'jetpack_amp_infinite_footers', array(), $buffer );
+
+		// Hook for themes to filter custom markup on next page load.
+		$buffer = apply_filters( 'jetpack_amp_infinite_output', $buffer );
+
+		// Add the amp next page markup.
+		$buffer = preg_replace(
+			'~</body>~',
+			$this->amp_get_footer_template( $footers ) . '$0',
+			$buffer
+		);
+
+		return $buffer;
+	}
+
+	/**
+	 * Get AMP next page markup with the custom footers.
+	 *
+	 * @param string[] $footers The theme footers.
+	 *
+	 * @return string
+	 */
+	protected function amp_get_footer_template( $footers = array() ) {
+		static $template = null;
```

Comment:
I’m a bit confused with the use of static  here as well as the null  check after setting it to null below.
Could you please elaborate on it?

Reply:
Hi @fgiannar 

Thanks for reviewing the code.

This method is supposed to be called twice in the runtime. The first time it runs happens when it's called from `amp_load_hooks()`. Because this is the first it runs, we declare a static `$template`. Then we check the contents for this variable, and if not set, we assign the contents of `amp_footer_template()`. In this method, we use output buffering to get a template for the footer. This first run warms the contents of the `$template`. 
The next time we run this method, it happens inside an output buffer callback. Because we declared the variable as being static, the `$template` now contains the footer markup so that the method will perform a slightly different action.
I tried a nested output buffer and have it done in one go, but I was having issues, likely because of the callback on the outer one.
This approach prevents having a property in this class whose only purpose is to be used by AMP workflow.

Please do let me know if I wasn't clear.

Reply:
Hey @pereirinha 

Many thanks for the explanation, it's clear now :)
Just a heads up here: The `$template` static variable will keep its state inside different instances of our class. If this is the desired behaviour, please feel free to resolve this.

Reply:
Yes, I'm aware of that, and it shouldn't be a problem.
I'm resolving. :)